metomi/rose/loc_handlers/git.py

```diff
@@ -0,0 +1,192 @@
+REC_COMMIT_HASH = re.compile(r"^[0-9a-f]+$")
+
+
+class GitLocHandler:
+    """Handler of Git locations."""
+
+    GIT = "git"
+    SCHEMES = [GIT]
+    WEB_SCHEMES = ["https"]
+    URI_SEPARATOR = "::"
+
+    def __init__(self, manager):
+        self.manager = manager
+        ret_code, versiontext, stderr = self.manager.popen.run(
+            "git", "version")
+        if ret_code:
+            self.git_version = None
+        else:
+            version_nums = []
+            for num_string in versiontext.split()[-1].split("."):
+                try:
+                    version_nums.append(int(num_string))
+                except ValueError:
+                    break
+            self.git_version = tuple(version_nums)
+
+    def can_pull(self, loc):
+        if self.git_version is None:
+            return False
+        scheme = urlparse(loc.name).scheme
+        if scheme in self.SCHEMES:
+            return True
+        if self.URI_SEPARATOR not in loc.name:
+            return False
+        remote = self._parse_name(loc)[0]
+        return (
+            scheme in self.WEB_SCHEMES
+            and not os.path.exists(loc.name)  # same as svn...
+            and not self.manager.popen.run(
+                "git", "ls-remote", "--exit-code", remote)[0]
+            # https://superuser.com/questions/227509/git-ping-check-if-remote-repository-exists
+        )
+
+    def parse(self, loc, conf_tree):
+        """Set loc.real_name, loc.scheme, loc.loc_type.
+
+        Within Git we have a lot of trouble figuring out remote
+        loc_type - a clone is required, unfortunately.
+
+        Short commit hashes will not work since there is no remote
+        rev-parse functionality. Long commit hashes will work if the
+        uploadpack.allowAnySHA1InWant configuration is set on the
+        remote repo or server.
+
+        Filtering requires uploadpack.allowFilter to be set true on
+        the remote repo or server.
+
+        """
+        loc.scheme = self.SCHEMES[0]
+        remote, path, ref = self._parse_name(loc)
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            git_dir_opt = f"--git-dir={tmpdirname}/.git"
+            self.manager.popen.run_ok(
+                "git", git_dir_opt, "init"
+            )
+            self.manager.popen.run_ok(
+                "git", git_dir_opt, "remote", "add", "origin", remote
+            )
+
+            # Make sure we configure for minimum fetching.
+            if self.git_version >= (2, 25, 0):
+                self.manager.popen.run_ok(
+                    "git", git_dir_opt, "sparse-checkout", "set", path,
+                    "--no-cone"
+                )
+            else:
+                self.manager.popen.run_ok(
+                    "git", git_dir_opt, "config", "extensions.partialClone",
+                    "true"
+                )
+
+            # Extract the commit hash if we don't already have it.
+            commithash = self._get_commithash(remote, ref)
```

Comment:
Very nice! :)

The comment was meant to say - given that ref can be a tag, or a branch, or a long commit hash or (forbidden for this implementation) a short commit hash, it could already be a long commit hash.